### PR TITLE
[IC3 ACS - Auth] <stable/2023-08-01>

### LINKED ISSUE #22473: [ACS Auth Service] API Review
New API Review meeting has been requested.

**Service Name**: ACS Auth Service
**Review Created By**: Maxim Rytych
**Review Date**: 3/8/2023 9:00 AM PT
**PR**: https://github.com/Azure/azure-rest-api-specs/pull/22306
**Hero Scenarios Link**: Not Provided
**Core Concepts Doc Link**: Not Provided

**Description**: Enable the possibility for the customers to use Identity APIs in an idempotent manner using new Repeatability-Request-ID and Repeatability-First-Sent headers.

Detailed meeting information and documents provided can be accessed [here](https://web.powerapps.com/apps/e41c69de-ef96-44ba-b287-0e012e7c13ae?view=edit&event=dfab315e-3ea6-ed11-aad1-000d3a341bd5)
For more information that will help prepare you for this review, the requirements, and office hours, visit the documentation [here](https://aka.ms/azsdk/onboarding/restapischedule)

### PR DESCRIPTION
**Pull request by maximrytych-ms**
**Changes in this PR enable the possibility for the customers to use Identity APIs in an idempotent manner using new Repeatability-Request-ID and Repeatability-First-Sent headers. **  
https://microsoft.sharepoint-df.com/:w:/t/IC3SDK/EaJumxZVaY5Ot-Rgb8oNt0MBviXc5BTfBrEeywmwkq1QuA

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [x] new API version
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  _August 2022_
  3. When do you expect to publish the swagger? Please provide date or, the date is not yet available, the month.
  _August 2022_
  4. If updating an existing version, please select the specific language SDKs and CLIs that must be refreshed after the swagger is published.
      - [x] SDK of .NET (need service team to ensure code readiness)
      - [x] SDK of Python
      - [x] SDK of Java
      - [x] SDK of Js

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

## Functionality description
This feature adds idempotent processing to the ACS Auth service according to [OASIS Repeatable Requests Version 1.0 specification](https://docs.oasis-open.org/odata/repeatable-requests/v1.0/repeatable-requests-v1.0.html). The service could avoid idempotency in the past because there were no significant negative side effects of calling the endpoints multiple times. Due to upcoming [Store custom attributes of ACS identities](https://microsoft.sharepoint-df.com/:w:/t/IC3SDK/ERFGZ2Jzb1pLlUigv8tV3rAB9hCnywdooNLbxzroWPi-rg?e=ReMhLA&wdOrigin=TEAMS-ELECTRON.p2p.bim&wdExp=TEAMS-TREATMENT&wdhostclicktime=1663324479661&wdLOR=cC998BC36-814C-4CFD-A4ED-F38C43912BF7) feature that will allow to store customer-defined data for identities the non-idempotent approach wouldn't be acceptable anymore, e.g. customers might get identities they either don't expect or store identities they don't have control over. There might be also the issue of storage space.

### Affected endpoints
Following endpoints will be affected by the change:

- POST {endpoint}/identities
- POST {endpoint}/identities/{id}/:revokeAccessTokens
- DELETE {endpoint}/identities/{id}
- POST {endpoint}/identities/{id}/:issueAccessToken
- POST {endpoint}/teamsUser/:exchangeAccessToken

Only the first 2 will be processing newly added headers beyond their validation. The rest of the endpoints is idempotent by its nature and will only validate the headers. These naturally idempotent headers are included to fulfill [Azure REST API guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#repeatability-of-requests) for the endpoints co-located behind a DNS.

image

### Old behavior
Previously it could happen that during the request for identity creation the identity got generated and stored in database, but due to network interrupt the user never got the originally created identity back. During retry new identity was created and this time returned back to the customer successfully, but the database contained additional "stale" identity that was outside of control of the customer and took additional space.
image

### New behavior
With the new behavior the request will contain 2 new headers: Repeatability-Request-Id and Repeatability-First-Sent. The first header will be used to identify the original request, the second one for repeatability window (5 minutes). If there is network interrupt, the retried request will check, whether there is a record with matching request id and within time window of 5 minutes, if yes the original identity is returned back to the customer. If not, the request is processed as usual. Using this approach the database will never contain "stale" identities.
image

## POST {endpoint}/identities/{id}/:revokeAccessTokens
image

### Old behavior
Previously it could happen that customer revoked tokens that shouldn't be revoked, e.g. customer wanted to revoke tokens for an identity and immediately afterwards issue a new token for that identity. The API for token revocation works in such a way that it calculates current time (DateTimeOffset.UtcNow), which is taken as a time before which all tokens for an identity should be revoked. If the revocation got interrupted due to network issues, customer meanwhile issued a new token for the identity and successful execution of retry began after the new token was issued, the new token was also revoked.

### New behavior
The request will be correlated based on the Repeatability-Request-Id and will use database cached value of RequestFirstReceived, which will be the current time of when the first request was received. That way possible retries will still contain the original time and on successful execution won't revoke fresh tokens.

## Review documentation
[Design document](https://microsoft.sharepoint-df.com/:w:/t/IC3SDK/EaJumxZVaY5Ot-Rgb8oNt0MBviXc5BTfBrEeywmwkq1QuA)

### SDK
There will be no changes to public API of  Azure/Communication/Identity SDKs except for the API version. The service layer will be regenerated based on the new swagger and convenience layer will be updated to automatically send the new headers with their appropriate values (new GUID and current date and time).

### ACS Auth service
Direct call to the service can be done using [this approach](https://learn.microsoft.com/en-us/azure/communication-services/tutorials/hmac-header-tutorial?pivots=programming-language-csharp). The code would just add new headers to the request as follows:

// generate new GUID and add it as a new request header
`var repeatabilityRequestId = Guid.NewGuid().ToString();`
`requestMessage.Headers.Add("Repeatability-Request-Id", repeatabilityRequestId);`

// generate current date according to IMF-fixdate form of HTTP-date as defined in [RFC7231] and add it as a new request header
`var repeatabilityFirstSent = DateTimeOffset.UtcNow.ToString("r", CultureInfo.InvariantCulture);`
`requestMessage.Headers.Add("Repeatability-First-Sent", repeatabilityFirstSent );`

## ~~Breaking Change Review Checklist~~ No breaking changes
If any of the following scenarios apply to the PR, request approval from the Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 
- [ ] Removing API(s) in a stable version
- [ ] Removing properties in a stable version
- [ ] Removing API version(s) in a stable version
- [ ] Updating API in a stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).
fix https://github.com/Azure/azure-rest-api-specs/issues/22473

### CODE REVIEW

File: specification/communication/data-plane/Identity/stable/2023-08-01/CommunicationIdentity.json [hunk 1]
```diff
@@ -0,0 +1,460 @@
...
+          }
+        }
+      }
+    },
+    "/identities/{id}": {
+      "delete": {
+        "tags": [
+          "Identity"
+        ],
+        "summary": "Delete the identity, revoke all tokens for the identity and delete all associated data.",
```
Comment (AikoBB) [inline specification/communication/data-plane/Identity/stable/2023-08-01/CommunicationIdentity.json:73]: there was a discussion about whether we should provide idempotency for the APIs that are idempotent naturally, I was wondering if we agreed to introduce the feature for all APIs?

File: specification/communication/data-plane/Identity/stable/2023-08-01/CommunicationIdentity.json [hunk 2]
```diff
@@ -0,0 +1,460 @@
...
+      "properties": {
+        "scopes": {
+          "description": "List of scopes attached to the token.",
+          "type": "array",
+          "items": {
+            "$ref": "#/definitions/CommunicationIdentityTokenScope"
+          }
+        },
+        "expiresInMinutes": {
+          "format": "int32",
```
Comment (AikoBB) [inline specification/communication/data-plane/Identity/stable/2023-08-01/CommunicationIdentity.json:349]: min, max, default fields are missing, same for line 380

File: specification/communication/data-plane/Identity/stable/2023-08-01/CommunicationIdentity.json [hunk 3]
```diff
@@ -0,0 +1,460 @@
...
+          "type": "integer"
+        }
+      }
+    },
+    "CommunicationIdentityTokenScope": {
+      "description": "List of scopes for an access token.",
+      "enum": [
+        "chat",
+        "voip",
+        "email"
```
Comment (AikoBB) [inline specification/communication/data-plane/Identity/stable/2023-08-01/CommunicationIdentity.json:392]: should we introduce this scope as well?

File: specification/communication/data-plane/Identity/stable/2023-08-01/CommunicationIdentity.json [hunk 4]
```diff
@@ -0,0 +1,460 @@
...
+      "required": [
+        "appId",
+        "token",
+        "userId"
+      ],
+      "type": "object",
+      "properties": {
+        "token": {
+          "description": "Azure AD access token of a Teams User to acquire a new Communication Identity access token.",
+          "minLength": 1,
```
Comment (AikoBB) [inline specification/communication/data-plane/Identity/stable/2023-08-01/CommunicationIdentity.json:410]: this field is newly introduced in this PR in a couple of places, does it really makes sense to introduce it?

File: specification/communication/data-plane/Identity/stable/2023-08-01/CommunicationIdentity.json [hunk 5]
```diff
@@ -0,0 +1,460 @@
...
+          "description": "List of scopes attached to the token.",
+          "type": "array",
+          "items": {
+            "$ref": "#/definitions/CommunicationIdentityTokenScope"
+          }
+        },
+        "expiresInMinutes": {
+          "format": "int32",
+          "description": "Optional custom validity period of the token within [60,1440] minutes range. If not provided, the default value of 1440 minutes (24 hours) will be used.",
+          "type": "integer"
```
Comment (AikoBB) [inline specification/communication/data-plane/Identity/stable/2023-08-01/CommunicationIdentity.json:351]: if we're specifying the format as int32, does it make sense to specify type as well

File: specification/communication/data-plane/Identity/stable/2023-08-01/examples/CreateIdentity.json [hunk 6]
```diff
@@ -0,0 +1,30 @@
+{
+  "parameters": {
+    "api-version": "2023-08-01",
+    "content-type": "application/json",
+    "endpoint": "https://my-resource.communication.azure.com",
+    "headers": {
+      "Repeatability-Request-ID": "4e7d453c-ae97-4778-ad51-cb3b4a701100",
+      "Repeatability-First-Sent": "2020-09-10T20:39:30.3244584+00:00"
```
Comment (AikoBB) [inline specification/communication/data-plane/Identity/stable/2023-08-01/examples/CreateIdentity.json:8]: maybe we should change the date after the date of the feature GA(2023-08-01)

File: specification/communication/data-plane/Identity/stable/2023-08-01/CommunicationIdentity.json [hunk 7]
```diff
@@ -0,0 +1,460 @@
+{
+  "swagger": "2.0",
+  "info": {
+    "title": "CommunicationIdentityClient",
+    "description": "Azure Communication Identity Service",
+    "version": "2023-08-01"
```
[the remark this answers is not in the document]
Comment (maximrytych-ms) [reply specification/communication/data-plane/Identity/stable/2023-08-01/CommunicationIdentity.json:6]: This date was specified in the design document: https://microsoft.sharepoint-df.com/:w:/t/IC3SDK/EaJumxZVaY5Ot-Rgb8oNt0MBviXc5BTfBrEeywmwkq1QuA.

File: specification/communication/data-plane/Identity/stable/2023-08-01/CommunicationIdentity.json (hunk 2 above)
Comment (maximrytych-ms) [reply specification/communication/data-plane/Identity/stable/2023-08-01/CommunicationIdentity.json:357]: Fixed.

File: specification/communication/data-plane/Identity/stable/2023-08-01/CommunicationIdentity.json [hunk 8]
```diff
@@ -0,0 +1,460 @@
...
+          "application/json"
+        ],
+        "produces": [
+          "application/json"
+        ],
+        "parameters": [
+          {
+            "in": "header",
+            "name": "Repeatability-Request-ID",
+            "description": "Unique identifier of the request.",
```
[the remark this answers is not in the document]
Comment (maximrytych-ms) [reply specification/communication/data-plane/Identity/stable/2023-08-01/CommunicationIdentity.json]: Fixed.

File: specification/communication/data-plane/Identity/stable/2023-08-01/CommunicationIdentity.json [hunk 9]
```diff
@@ -0,0 +1,460 @@
...
+            "type": "string"
+          },
+          {
+            "in": "header",
+            "name": "Repeatability-First-Sent",
+            "description": "Date and time at which the request was first created.",
```
[the remark this answers is not in the document]
Comment (maximrytych-ms) [reply specification/communication/data-plane/Identity/stable/2023-08-01/CommunicationIdentity.json]: Fixed.

File: specification/communication/data-plane/Identity/stable/2023-08-01/CommunicationIdentity.json [hunk 10]
```diff
@@ -0,0 +1,460 @@
...
+            "schema": {
+              "$ref": "#/definitions/CommunicationIdentityCreateRequest"
+            }
+          }
+        ],
+        "responses": {
+          "default": {
+            "description": "Error",
+            "schema": {
+              "$ref": "../../../Common/stable/2021-03-07/common.json#/definitions/CommunicationErrorResponse"
```
[the remark this answers is not in the document]
Comment (maximrytych-ms) [reply specification/communication/data-plane/Identity/stable/2023-08-01/CommunicationIdentity.json]: Updated.

File: specification/communication/data-plane/Identity/stable/2023-08-01/CommunicationIdentity.json (hunk 3 above)
Comment (maximrytych-ms) [reply specification/communication/data-plane/Identity/stable/2023-08-01/CommunicationIdentity.json]: Removed.

File: specification/communication/data-plane/Identity/stable/2023-08-01/CommunicationIdentity.json (hunk 4 above)
Comment (maximrytych-ms) [reply specification/communication/data-plane/Identity/stable/2023-08-01/CommunicationIdentity.json]: Removed.

File: specification/communication/data-plane/Identity/stable/2023-08-01/CommunicationIdentity.json (hunk 5 above)
Comment (maximrytych-ms) [reply specification/communication/data-plane/Identity/stable/2023-08-01/CommunicationIdentity.json]: I kept is, as it's present in previous version as well and my changes are not related.

File: specification/communication/data-plane/Identity/stable/2023-08-01/examples/CreateIdentity.json (hunk 6 above)
Comment (maximrytych-ms) [reply specification/communication/data-plane/Identity/stable/2023-08-01/examples/CreateIdentity.json]: Good point. Fixed.

File: specification/communication/data-plane/Identity/stable/2023-08-01/CommunicationIdentity.json (hunk 1 above)
Comment (AikoBB) [reply specification/communication/data-plane/Identity/stable/2023-08-01/CommunicationIdentity.json:75]: I guess for this question @petrsvihlik can answer :)

Comment (petrsvihlik) [reply specification/communication/data-plane/Identity/stable/2023-08-01/CommunicationIdentity.json:75]: We'll just make these APIs aware of the repeatability headers and keep the current functionality, following this Azure [guideline](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#repeatability-of-requests):

> When understood, all endpoints co-located behind a DNS name MUST understand the header. This means that a service MUST NOT ignore the presence of a header for any endpoints behind the DNS name, but rather fail the request containing a Repeatability-Request-ID header if that particular endpoint lacks support for repeatable requests. Such partial support SHOULD be avoided due to the confusion it causes for clients.


That means when someone decides to set the headers, they must be set correctly - we'll perform the header validation and continue with the normal, naturally idempotent execution.

File: specification/communication/data-plane/Identity/stable/2023-08-01/CommunicationIdentity.json [hunk 11]
```diff
@@ -0,0 +1,470 @@
...
+            "required": true,
+            "type": "string"
+          },
+          {
+            "$ref": "#/parameters/ApiVersionParameter"
+          },
+          {
+            "in": "header",
+            "name": "Repeatability-Request-ID",
+            "description": "If specified, the client directs that the request is repeatable; that is, that the client can make the request multiple times with the same Repeatability-Request-Id and get back an appropriate response without the server executing the request multiple times. The value of the Repeatability-Request-Id  is an opaque string representing a client-generated, globally unique for all time, identifier for the request. It is recommended to use version 4 (random) UUIDs.",
```
Comment (petrsvihlik) [inline specification/communication/data-plane/Identity/stable/2023-08-01/CommunicationIdentity.json:267]:
```suggestion
            "description": "If specified, the client directs that the request is repeatable; that is, that the client can make the request multiple times with the same Repeatability-Request-Id and get back an appropriate response without the server executing the request multiple times. The value of the Repeatability-Request-Id is an opaque string representing a client-generated, globally unique for all time, identifier for the request. It is recommended to use version 4 (random) UUIDs.",
```

Comment (petrsvihlik) [reply specification/communication/data-plane/Identity/stable/2023-08-01/CommunicationIdentity.json:267]: can you also highlight the fact that Contoso should not use any of their internal identifiers when issuing the idempotency identifier (Repeatability-Request-ID). it really needs to be an opaque and meaningless string.

**Review by petrsvihlik — CHANGES_REQUESTED** [review]
just one nit...otherwise, looks fine

File: specification/communication/data-plane/Identity/stable/2023-08-01/CommunicationIdentity.json (hunk 11 above)
Comment (maximrytych-ms) [reply specification/communication/data-plane/Identity/stable/2023-08-01/CommunicationIdentity.json]: Ok. I added the formulation.

**Review by petrsvihlik — APPROVED** [review]

File: specification/communication/data-plane/Identity/stable/2023-08-01/examples/CreateIdentity.json [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+          "id": "8:acs:2dee53b4-368b-45b4-ab52-8493fb117652_00000005-14a2-493b-8a72-5a3a0d000081"
+        },
+        "accessToken": {
+          "token": "token",
+          "expiresOn": "2023-09-10T21:39:39.3244584+00:00"
+        }
+      }
+    }
+  }
+}
```
Comment (petrsvihlik) [inline specification/communication/data-plane/Identity/stable/2023-08-01/examples/CreateIdentity.json:30]: also, add the `Repeatability-Result` pls (similarly to [email](https://github.com/Azure/azure-rest-api-specs/search?q=%22Repeatability-Result%22)). 

btw - is it possible to add it also to the swagger, not only the examples?

**Review by petrsvihlik — CHANGES_REQUESTED** [review]

File: specification/communication/data-plane/Identity/stable/2023-08-01/examples/CreateIdentity.json (hunk 12 above)
Comment (maximrytych-ms) [reply specification/communication/data-plane/Identity/stable/2023-08-01/examples/CreateIdentity.json]: Done.

**Review by petrsvihlik — APPROVED** [review]

**Review by AikoBB — APPROVED** [review]

**Review by mikekistler — APPROVED** [review]
Looks good. 👍

**Comment by petrsvihlik** [comment]
@jhendrixMSFT can you please help us get this merged?

**Pull request merged**